packages/react-components/src/data/transformers/anomaly/arrow/transformer.ts

```diff
@@ -0,0 +1,215 @@
+    const isArrowDataSource = super.canTransform(datasource);
+    if (!isArrowDataSource) return false;
+
+    const hasValidData = this.#validData(datasource);
+
+    if (!hasValidData) {
+      console.warn(
+        'Anomaly arrow datasource has invalid data shape.',
+        datasource
+      );
+    }
+
+    return hasValidData;
+  }
+
+  #getDataExtent(dataSource: AnomalyArrowDataSource) {
+    const { fields } = dataSource.value.data;
+    const timestampField = this.#timestampField(fields);
+
+    if (!timestampField) return undefined;
+
+    let start: Date | null = null;
+    let end: Date | null = null;
+    (timestampField.values as number[]).forEach((timestamp) => {
+      const date = new Date(timestamp);
+      if (start === null || isBefore(date, start)) {
+        start = date;
+      }
+      if (end === null || isAfter(date, end)) {
+        end = date;
+      }
+    });
+
+    if (start === null || end === null) return undefined;
+
+    return {
+      start,
+      end,
+    };
+  }
+
+  #diagnosticId(order: number): `diagnostic_${string}` {
+    return `diagnostic_${order}`;
+  }
+
+  #getDiagnosticFieldMapping(fields: Field[]) {
+    return fields
+      .filter(isDiagnosticField)
+      .map((field) => field.name)
+      .sort()
+      .reduce((acc, n, i) => {
+        acc[n] = this.#diagnosticId(i);
+        return acc;
+      }, {} as DiagnosticNameMapping);
+  }
+
+  #timestampField(fields: Field[]) {
+    return fields.find((field) => field.name === TIMESTAMP_FIELD_NAME);
+  }
+
+  #diagnosticField(fields: Field[], name: string) {
+    return fields.find((field) => field.name === name);
+  }
+
+  #predictionReasonField(fields: Field[]) {
+    return fields.find((field) => field.name === PREDICTION_REASON_FIELD_NAME);
+  }
+
+  transform(dataSource: AnomalyArrowDataSource): AnomalyData {
+    const { fields } = dataSource.value.data;
+
+    const emptyData = { timestamp: [] };
+
+    const predictionReasonField = this.#predictionReasonField(fields);
+    const timestampField = this.#timestampField(fields);
+
+    if (!predictionReasonField || !timestampField) return emptyData;
+
+    const diagnosticFieldMapping = this.#getDiagnosticFieldMapping(fields);
+
+    /**
+     * echarts dataset can be created in a keyed column format,
+     * but this data format is not supported in the out of
+     * the box echarts dataset sort transform. In order to not
+     * write our own custom sort transform, we are transforming
+     * the data into an object array format.
+     *
+     * https://github.com/apache/echarts/blob/a2824713ac98378f4effc0baa372798c84cc4127/src/component/transform/sortTransform.ts#L51
+     */
+    const data: AnomalyData = [];
+
+    let i = 0;
+    while (i < predictionReasonField.values.length) {
+      // filter out values that are not anomalous
+      if (!isAnomalousValue(predictionReasonField.values.at(i) as string)) {
+        fields.forEach((field) => {
+          field.values.splice(i, 1);
```

Comment:
is the `field` safe to modify in place? would the effect surface to integrating app?